### PR TITLE
Error reasons integration test

### LINKED ISSUE #1609: Integration test for error reasons
1. Create group with an invalid key name, because autoscale does not
   verify the key and hence won't just 400 the create group request.
2. Assert group goes into error state with the reason being invalid
   Server config.
3. Delete the key, and converge.
4. Assert group goes into active state.
5. Change the launch config to have an invalid load balancer.
6. Scale up.
7. Assert group goes into error state with the reason being an invalid
   CLB.

### PR DESCRIPTION
**Pull request by cyli**
Fixes #1609.

This is dependent upon #1608 to merge for this test to pass.

### CODE REVIEW

File: otter/integration/tests/test_convergence.py [hunk 1]
```diff
@@ -989,6 +990,101 @@ def test_recover_from_identity_auth_failures(self):
             max_entities=10)
         yield self.helper.start_group_and_wait(group, rcs, desired=5)
 
+    @inlineCallbacks
+    def test_error_reasons_are_updated(self):
```
Comment (manishtomar) [inline otter/integration/tests/test_convergence.py]: How about skipping this with link to #1578 and merging it? Then I can merge #1608 with this enabled. That way we that merge will be synchronous with the test.

File: otter/convergence/errors.py [hunk 2]
```diff
@@ -43,6 +47,11 @@ def _present_clb_deleted_error(exception):
             % (exception.lb_id,))
 
 
+@_present_exception.register(CreateServerConfigurationError)
+def _present_server_configuration_error(exception):
+    return "Server launch configuration is invalid: %s" % (exception.message,)
```
Comment (manishtomar) [inline otter/convergence/errors.py]: Non-blocking: How come not using `format`?

Comment (cyli) [inline otter/convergence/errors.py]: I was mostly just following the previous convention of the other `_present_server_configuration_error` things.  I generally prefer `format`, but I wasn't sure if we still follow that.  I'll be happy to change both this and all the rest to use format instead :)

File: otter/integration/tests/test_convergence.py (hunk 1 above)
Comment (cyli) [inline otter/integration/tests/test_convergence.py]: Sounds good!

File: otter/integration/tests/test_convergence.py [hunk 3]
```diff
@@ -989,6 +990,101 @@ def test_recover_from_identity_auth_failures(self):
...
+            min_entities=1, max_entities=10,
+            server_name_prefix="no-such-key",
+            key_name="invalid_key_name"
+        )
+        launch_config = deepcopy(group.group_config['launchConfiguration'])
+
+        if not not_mimic():
+            # if this is mimic, we have to make server creation fail.
+            # in production, it's actually an invalid key and hence would
+            # naturally fail.
```
Comment (manishtomar) [inline otter/integration/tests/test_convergence.py]: Non-blocking thought: I wonder if mimic should error in these situations. Does it error on invalid image currently? I always thought mimic to simulate behavior that is not possible through _normal_ APIs. Maybe create ticket in mimic to replicate this?

File: otter/integration/tests/test_convergence.py [hunk 4]
```diff
@@ -989,6 +990,101 @@ def test_recover_from_identity_auth_failures(self):
...
+        yield group.wait_for_state(
+            self.rcs,
+            MatchesAll(
+                ContainsDict({
+                    'pendingCapacity': Equals(0),
+                    'desiredCapacity': Equals(1),
+                    'active': HasLength(1),
+                    'status': Equals("ACTIVE"),
+                }),
+                Not(Contains('error'))
```
Comment (manishtomar) [inline otter/integration/tests/test_convergence.py]: Did you mean `errors`?

**Comment by manishtomar** [comment]
LGTM modulo comments :)

File: otter/integration/tests/test_convergence.py (hunk 4 above)
Comment (cyli) [inline otter/integration/tests/test_convergence.py]: Doh!  yes, good catch, thanks!

File: otter/integration/tests/test_convergence.py (hunk 3 above)
Comment (cyli) [inline otter/integration/tests/test_convergence.py]: Mimic doesn't actually implement glance or the key store or anything.  It has some preset invalid images that you can use to fail the server, but it doesn't  have a set of preset keys.  We should probably add that, yeah.

Comment (cyli) [inline otter/integration/tests/test_convergence.py]: https://github.com/rackerlabs/mimic/issues/346

**Comment by cyli** [comment]
jenkins please run integration tests

**Comment by cyli** [comment]
Merging, thanks @manishtomar!

**Pull request merged**